subsys/bluetooth/services/ias/ias_client.c

```diff
@@ -0,0 +1,241 @@
+#define LOG_MODULE_NAME bt_ias_client
+#include "../../common/log.h"
+
+enum {
+	IAS_DISCOVER_IN_PROGRSS,
+
+	IAS_NUM_FLAGS, /* keep as last */
+};
+
+struct bt_ias_client {
+	/* Handle for alert writes */
+	uint16_t write_handle;
+
+	/** Internal flags **/
+	ATOMIC_DEFINE(flags, IAS_NUM_FLAGS);
+
+	/* Gatt discover procedure parameters */
+	struct bt_gatt_discover_params discover;
+};
+
+static const struct bt_uuid *alert_lvl_uuid = BT_UUID_ALERT_LEVEL;
+static const struct bt_uuid *ias_uuid = BT_UUID_IAS;
+
+static const struct bt_ias_client_cb *ias_client_cb;
+static struct bt_ias_client client_list[CONFIG_BT_MAX_CONN];
+
+static struct bt_ias_client *client_by_conn(struct bt_conn *conn)
+{
+	return &client_list[bt_conn_index(conn)];
+}
+
+static void client_cleanup(struct bt_ias_client *ias_client)
+{
+	(void)memset(ias_client, 0, sizeof(*ias_client));
+}
+
+static void discover_failed(struct bt_conn *conn, int err)
+{
+	BT_DBG("conn %p", (void *)conn);
+
+	client_cleanup(client_by_conn(conn));
+	ias_client_cb->discover(conn, err);
+}
+
+int bt_ias_client_alert_high(struct bt_conn *conn)
+{
+	int err;
+	const uint8_t lvl = BT_IAS_ALERT_LVL_HIGH_ALERT;
+
+	CHECKIF(conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	if (client_by_conn(conn)->write_handle == 0) {
+		return BT_ATT_ERR_INVALID_HANDLE;
+	}
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write_handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client high alert level write failed: %d", err);
+
+		return err;
+	}
+
+	return err;
+}
+
+int bt_ias_client_alert_mild(struct bt_conn *conn)
+{
+	int err;
+	const uint8_t lvl = BT_IAS_ALERT_LVL_MILD_ALERT;
+
+	CHECKIF(conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	if (client_by_conn(conn)->write_handle == 0) {
+		return BT_ATT_ERR_INVALID_HANDLE;
+	}
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write_handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client mild alert level write failed: %d", err);
+
+		return err;
+	}
+
+	return err;
+}
+
+int bt_ias_client_alert_stop(struct bt_conn *conn)
+{
+	int err;
+	const uint8_t lvl = BT_IAS_ALERT_LVL_NO_ALERT;
+
+	CHECKIF(conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	if (client_by_conn(conn)->write_handle == 0) {
+		return BT_ATT_ERR_INVALID_HANDLE;
+	}
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write_handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client stop alert write failed: %d", err);
+
+		return err;
+	}
+
+	return err;
+}
+
+static uint8_t bt_ias_alert_lvl_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	const struct bt_gatt_chrc *chrc = (struct bt_gatt_chrc *)attr->user_data;
+
+	atomic_clear_bit(client_by_conn(conn)->flags, IAS_DISCOVER_IN_PROGRSS);
+
+	if (attr == NULL) {
+		ias_client_cb->discover(conn, -BT_ATT_ERR_ATTRIBUTE_NOT_FOUND);
```

Comment:
```suggestion
		discover_failed(conn, -ENOENT);
```

because `bt_ias_client_cb.discover` API says `err 0 on success, ATT error or negative errno otherwise`